abci/types/types.go

```diff
@@ -167,3 +170,125 @@ func RespondVerifyVoteExtension(ok bool) ResponseVerifyVoteExtension {
 		Result: result,
 	}
 }
+
+func MustHashResults(r []*ExecTxResult) []byte {
+	return merkle.HashFromByteSlices(mustResultsToByteSlices(r))
+}
+
+// ProveResult returns a merkle proof of one result from the set
+func MustProveResult(r []*ExecTxResult, i int) merkle.Proof {
+	_, proofs := merkle.ProofsFromByteSlices(mustResultsToByteSlices(r))
+	return *proofs[i]
+}
+
+// deterministicExecTxResult strips non-deterministic fields from
+// ResponseDeliverTx and returns another ResponseDeliverTx.
+func deterministicExecTxResult(response *ExecTxResult) *ExecTxResult {
+	return &ExecTxResult{
+		Code:      response.Code,
+		Data:      response.Data,
+		GasWanted: response.GasWanted,
+		GasUsed:   response.GasUsed,
+	}
+}
+
+func mustResultsToByteSlices(r []*ExecTxResult) [][]byte {
+	s := make([][]byte, len(r))
+	for i, e := range r {
+		d := deterministicExecTxResult(e)
+		b, err := d.Marshal()
+		if err != nil {
+			panic(err)
+		}
+		s[i] = b
+	}
+	return s
+}
+
+func (tr *TxRecord) IsIncluded() bool {
+	return tr.Action == TxRecord_ADDED || tr.Action == TxRecord_UNMODIFIED
+}
+
+// IncludedTxs returns all of the TxRecords that are marked for inclusion in the
+// proposed block.
+func (rpp *ResponsePrepareProposal) IncludedTxs() []*TxRecord {
+	trs := []*TxRecord{}
+	for _, tr := range rpp.TxRecords {
+		if tr.IsIncluded() {
+			trs = append(trs, tr)
+		}
+	}
+	return trs
+}
+
+// RemovedTxs returns all of the TxRecords that are marked for removal from the
+// mempool.
+func (rpp *ResponsePrepareProposal) RemovedTxs() []*TxRecord {
+	var trs []*TxRecord
+	for _, tr := range rpp.TxRecords {
+		if tr.Action == TxRecord_REMOVED {
+			trs = append(trs, tr)
+		}
+	}
+	return trs
+}
+
+// AddedTxs returns all of the TxRecords that are marked as added to the proposal.
+func (rpp *ResponsePrepareProposal) AddedTxs() []*TxRecord {
+	var trs []*TxRecord
+	for _, tr := range rpp.TxRecords {
+		if tr.Action == TxRecord_ADDED {
+			trs = append(trs, tr)
+		}
+	}
+	return trs
+}
+
+// Validate checks that the fields of the ResponsePrepareProposal are properly
+// constructed. Validate returns an error if any of the validation checks fail.
+func (rpp *ResponsePrepareProposal) Validate(maxSizeBytes int64, otxs [][]byte) error {
+
+	if !rpp.ModifiedTx {
+		// This method currently only checks the validity of the TxRecords field.
+		// If ModifiedTx is false, then we can ignore the validity of the TxRecords field.
+		//
+		// TODO: When implementing VoteExensions, AppSignedUpdates may be modified by the application
```

Comment:
Wow! Good catch! This is a leftover I didn't spot when updating the spec to the new design for vote extensions. Sorry about that.
Please remove it if you can, otherwise I'll remove it next week. If you manage to remove it in this PR, please also remove the corresponding field in the table in `abci++_methods_002_draft.md`, as well as the two bullet points in "Usage" below the table that mention `app_signed_updates`